openedx/core/lib/call_stack_manager.py

```diff
@@ -0,0 +1,120 @@
+"""
+Get call stacks of Model Class
+in three cases-
+1. QuerySet API
+2. save()
+3. delete()
+
+classes:
+Globals - contains global stack_book dictionary , also regular expression filters
+CallStackManager -  stores all stacks in global dictionary and logs
+
+How to use-
+1. Import following in the file where class to be tracked resides
+    from openedx.core.lib.call_stack_manager import CallStackManager
+2. Override objects of default manager by writing following in any model class which you want to track-
+    objects = CallStackManager()
+
+Note -
+1.Format for stack_book
+{
+    "modelclass1":
+        [[(frame 1),(frame 2)],
+         [(frame 11),(frame21)]]
+    "modelclass2":
+        [[(frame 3),(frame 4)],
+         [(frame 6),(frame 5)]]
+
+}
+where frame is a tuple of
+(file path, Line Number, Context)
+"""
+
+import logging
+import traceback
+import re
+import threading
+
+from django.db.models import Manager
+
+
+log = logging.getLogger(__name__)
+
+
+class Globals(threading.local):
+    """
+    stores necessary global data structures for Call Stacks and runs a method necessary for
+    """
+    # dictionary which stores call stacks.
+    # { "ModelClasses" : [ListOfFrames]}
+    # Frames - ('FilePath','LineNumber','Context')
+    # ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
+    #                                                     [(file,line number,context),(---,---,---)]]}
+    stack_book = {}
+
+    # filter to trickle down call stacks.
+    exclude = ['^.*python2.7.*$', '^.*call_stack_manager.*$']
+    regular_expressions = [re.compile(x) for x in exclude]
```

Comment:
I think these should probably all be module-level, rather than part of a thread-local object. The constants certainly don't, but also, if you put the `stack_book` in a thread-local, then you'll log the same stack trace in each thread that encounters it, rather than only in the first thread to do so.